File: frontend-src/src-tauri/src/services/programming_units.rs
use rusqlite::{params, Connection, Row};
use serde_json::{json, Value};

use crate::db;
use crate::error::ApiError;

use super::merge_object;

const COLUMNS: &str = "id, course_id, name, sessions, start_date, session_details, \
    linked_criteria_ids, linked_basic_knowledge_ids, created_at, updated_at, \
    linked_specific_competence_ids, context, final_product, final_exam";

fn id_array_to_json(raw: &str) -> Value {
    serde_json::from_str::<Value>(raw).unwrap_or_else(|_| json!([]))
}

fn row_to_json(row: &Row) -> rusqlite::Result<Value> {
    let session_details: String = row.get(5)?;
    let linked_criteria_ids: String = row.get(6)?;
    let linked_basic_knowledge_ids: String = row.get(7)?;
    let linked_specific_competence_ids: String = row.get(10)?;
    let final_product: String = row.get(12)?;
    let final_exam: String = row.get(13)?;
    Ok(json!({
        "id": row.get::<_, String>(0)?,
        "courseId": row.get::<_, String>(1)?,
        "name": row.get::<_, String>(2)?,
        "sessions": row.get::<_, i64>(3)?,
        "startDate": row.get::<_, Option<String>>(4)?,
        "sessionDetails": id_array_to_json(&session_details),
        "linkedCriteriaIds": id_array_to_json(&linked_criteria_ids),
        "linkedBasicKnowledgeIds": id_array_to_json(&linked_basic_knowledge_ids),
        "createdAt": row.get::<_, String>(8)?,
        "updatedAt": row.get::<_, String>(9)?,
        "linkedSpecificCompetenceIds": id_array_to_json(&linked_specific_competence_ids),
        "context": row.get::<_, Option<String>>(11)?,
        "finalProduct": serde_json::from_str::<Value>(&final_product).unwrap_or_else(|_| json!({"incluido": false})),
        "finalExam": serde_json::from_str::<Value>(&final_exam).unwrap_or_else(|_| json!({"incluido": false})),
    }))
}

fn get_one(conn: &Connection, id: &str) -> Result<Option<Value>, ApiError> {
    let mut stmt = conn.prepare(&format!("SELECT {COLUMNS} FROM programming_units WHERE id = ?"))?;
    let mut rows = stmt.query_map(params![id], row_to_json)?;
    match rows.next() {
        Some(row) => Ok(Some(row?)),
        None => Ok(None),
    }
}

pub fn list(conn: &Connection, course_id: &str) -> Result<Value, ApiError> {
    let mut stmt = conn.prepare(&format!(
        "SELECT {COLUMNS} FROM programming_units WHERE course_id = ? ORDER BY start_date IS NULL, start_date, name"
    ))?;
    let rows = stmt.query_map(params![course_id], row_to_json)?;
    let items: Result<Vec<Value>, _> = rows.collect();
    Ok(Value::Array(items?))
}

pub fn create(conn: &Connection, course_id: &str, body: Value) -> Result<Value, ApiError> {
    let name = body.get("name").and_then(Value::as_str)
        .ok_or_else(|| ApiError::bad_request("name es obligatorio"))?;
    let sessions = body.get("sessions").and_then(Value::as_i64).unwrap_or(0);
    let start_date = body.get("startDate").and_then(Value::as_str);
    let session_details = body.get("sessionDetails").cloned().unwrap_or_else(|| json!([]));
    let linked_criteria = body.get("linkedCriteriaIds").cloned().unwrap_or_else(|| json!([]));
    let linked_knowledge = body.get("linkedBasicKnowledgeIds").cloned().unwrap_or_else(|| json!([]));
    let linked_specific_competences = body.get("linkedSpecificCompetenceIds").cloned().unwrap_or_else(|| json!([]));
    let context = body.get("context").and_then(Value::as_str);
    let final_product = body.get("finalProduct").cloned().unwrap_or_else(|| json!({"incluido": false}));
    let final_exam = body.get("finalExam").cloned().unwrap_or_else(|| json!({"incluido": false}));

    let id = db::new_uuid();
    let now = db::now_iso();
    conn.execute(
        "INSERT INTO programming_units (id, course_id, name, sessions, start_date, session_details, linked_criteria_ids, linked_basic_knowledge_ids, created_at, updated_at, linked_specific_competence_ids, context, final_product, final_exam) \
         VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?)",
        params![
            id, course_id, name, sessions, start_date,
            serde_json::to_string(&session_details).map_err(ApiError::internal)?,
            serde_json::to_string(&linked_criteria).map_err(ApiError::internal)?,
            serde_json::to_string(&linked_knowledge).map_err(ApiError::internal)?,
            now.clone(), now,
            serde_json::to_string(&linked_specific_competences).map_err(ApiError::internal)?,
            context,
            serde_json::to_string(&final_product).map_err(ApiError::internal)?,
            serde_json::to_string(&final_exam).map_err(ApiError::internal)?,
        ],
    )?;
    get_one(conn, &id)?.ok_or_else(|| ApiError::internal("no se pudo releer la unidad de programación recién creada"))
}

pub fn update(conn: &Connection, id: &str, body: Value) -> Result<Value, ApiError> {
    let current = get_one(conn, id)?.ok_or_else(|| ApiError::not_found("Unidad de programación no encontrada."))?;
    let merged = merge_object(&current, &body);
    let name = merged.get("name").and_then(Value::as_str).unwrap_or_default();
    let sessions = merged.get("sessions").and_then(Value::as_i64).unwrap_or(0);
    let start_date = merged.get("startDate").and_then(Value::as_str);
    let session_details = merged.get("sessionDetails").cloned().unwrap_or_else(|| json!([]));
    let linked_criteria = merged.get("linkedCriteriaIds").cloned().unwrap_or_else(|| json!([]));
    let linked_knowledge = merged.get("linkedBasicKnowledgeIds").cloned().unwrap_or_else(|| json!([]));
    let linked_specific_competences = merged.get("linkedSpecificCompetenceIds").cloned().unwrap_or_else(|| json!([]));
    let context = merged.get("context").and_then(Value::as_str);
    let final_product = merged.get("finalProduct").cloned().unwrap_or_else(|| json!({"incluido": false}));
    let final_exam = merged.get("finalExam").cloned().unwrap_or_else(|| json!({"incluido": false}));

    conn.execute(
        "UPDATE programming_units SET name = ?, sessions = ?, start_date = ?, session_details = ?, linked_criteria_ids = ?, linked_basic_knowledge_ids = ?, updated_at = ?, linked_specific_competence_ids = ?, context = ?, final_product = ?, final_exam = ? WHERE id = ?",
        params![
            name, sessions, start_date,
            serde_json::to_string(&session_details).map_err(ApiError::internal)?,
            serde_json::to_string(&linked_criteria).map_err(ApiError::internal)?,
            serde_json::to_string(&linked_knowledge).map_err(ApiError::internal)?,
            db::now_iso(),
            serde_json::to_string(&linked_specific_competences).map_err(ApiError::internal)?,
            context,
            serde_json::to_string(&final_product).map_err(ApiError::internal)?,
            serde_json::to_string(&final_exam).map_err(ApiError::internal)?,
            id,
        ],
    )?;
    get_one(conn, id)?.ok_or_else(|| ApiError::internal("no se pudo releer la unidad de programación tras actualizar"))
}

pub fn delete(conn: &Connection, id: &str) -> Result<Value, ApiError> {
    let changed = conn.execute("DELETE FROM programming_units WHERE id = ?", params![id])?;
    if changed == 0 {
        return Err(ApiError::not_found("Unidad de programación no encontrada."));
    }
    Ok(Value::Null)
}

File: api/app/services/programming_units.py
import uuid
from datetime import date
from typing import Optional

from psycopg.types.json import Json

from services.db import get_conn
from services.schemas import ApiModel

_COLUMNS = """
    id, course_id, name, sessions, start_date, context, session_details,
    linked_criteria_ids, linked_basic_knowledge_ids, linked_specific_competence_ids,
    final_product, final_exam, created_at, updated_at
"""

_JSON_FIELDS = {"session_details", "final_product", "final_exam"}


class ProgrammingUnitInput(ApiModel):
    name: str
    sessions: int = 0
    start_date: Optional[date] = None
    context: Optional[str] = None
    session_details: list = []
    linked_criteria_ids: list[uuid.UUID] = []
    linked_basic_knowledge_ids: list[uuid.UUID] = []
    linked_specific_competence_ids: list[uuid.UUID] = []
    final_product: dict = {"incluido": False}
    final_exam: dict = {"incluido": False}


class ProgrammingUnitPatch(ApiModel):
    name: Optional[str] = None
    sessions: Optional[int] = None
    start_date: Optional[date] = None
    context: Optional[str] = None
    session_details: Optional[list] = None
    linked_criteria_ids: Optional[list[uuid.UUID]] = None
    linked_basic_knowledge_ids: Optional[list[uuid.UUID]] = None
    linked_specific_competence_ids: Optional[list[uuid.UUID]] = None
    final_product: Optional[dict] = None
    final_exam: Optional[dict] = None


class ProgrammingUnit(ProgrammingUnitInput):
    id: uuid.UUID
    course_id: uuid.UUID


def list_programming_units(course_id: str) -> list[ProgrammingUnit]:

    with get_conn() as conn:

        with conn.cursor() as cur:

            cur.execute(f"SELECT {_COLUMNS} FROM programming_units WHERE course_id = %s ORDER BY start_date NULLS LAST, name", [course_id])

            return [ProgrammingUnit.model_validate(row) for row in cur.fetchall()]


def create_programming_unit(course_id: str, data: ProgrammingUnitInput) -> ProgrammingUnit:

    with get_conn() as conn:

        with conn.cursor() as cur:

            cur.execute(
                f"""
                INSERT INTO programming_units
                    (course_id, name, sessions, start_date, context, session_details,
                     linked_criteria_ids, linked_basic_knowledge_ids, linked_specific_competence_ids,
                     final_product, final_exam)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s) RETURNING {_COLUMNS}
                """,
                [
                    course_id, data.name, data.sessions, data.start_date, data.context, Json(data.session_details),
                    [str(i) for i in data.linked_criteria_ids], [str(i) for i in data.linked_basic_knowledge_ids],
                    [str(i) for i in data.linked_specific_competence_ids],
                    Json(data.final_product), Json(data.final_exam),
                ]
            )

            return ProgrammingUnit.model_validate(cur.fetchone())


def update_programming_unit(unit_id: str, data: ProgrammingUnitPatch) -> Optional[ProgrammingUnit]:

    fields = data.model_dump(exclude_unset=True)

    processed = {}

    for key, value in fields.items():

        if key in _JSON_FIELDS:
            processed[key] = Json(value)
        elif key in ("linked_criteria_ids", "linked_basic_knowledge_ids", "linked_specific_competence_ids"):
            processed[key] = [str(i) for i in value]
        else:
            processed[key] = value

    with get_conn() as conn:

        with conn.cursor() as cur:

            if processed:

                set_clause = ", ".join(f"{key} = %s" for key in processed)

                cur.execute(
                    f"UPDATE programming_units SET {set_clause}, updated_at = now() WHERE id = %s RETURNING {_COLUMNS}",
                    [*processed.values(), unit_id]
                )

                row = cur.fetchone()

            else:

                cur.execute(f"SELECT {_COLUMNS} FROM programming_units WHERE id = %s", [unit_id])

                row = cur.fetchone()

            return ProgrammingUnit.model_validate(row) if row else None


def delete_programming_unit(unit_id: str) -> bool:

    with get_conn() as conn:

        with conn.cursor() as cur:

            cur.execute("DELETE FROM programming_units WHERE id = %s", [unit_id])

            return cur.rowcount > 0
